frame/staking/src/pallet/impls.rs

```diff
@@ -629,54 +639,79 @@ impl<T: Config> Pallet<T> {
 
 	/// Get all of the voters that are eligible for the npos election.
 	///
-	/// This will use all on-chain nominators, and all the validators will inject a self vote.
+	/// `voter_count` imposes a cap on the number of voters returned; care should be taken to ensure
+	/// that it is accurate.
+	///
+	/// This will use on-chain nominators, and all the validators will inject a self vote.
 	///
 	/// This function is self-weighing as [`DispatchClass::Mandatory`].
 	///
 	/// ### Slashing
 	///
 	/// All nominations that have been submitted before the last non-zero slash of the validator are
 	/// auto-chilled.
-	pub fn get_npos_voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		let weight_of = Self::slashable_balance_of_fn();
-		let mut all_voters = Vec::new();
+	pub fn get_npos_voters(
+		maybe_max_len: Option<usize>,
+	) -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
+		let nominator_count = CounterForNominators::<T>::get() as usize;
+		let validator_count = CounterForValidators::<T>::get() as usize;
+		let all_voter_count = validator_count.saturating_add(nominator_count);
+		drop(validator_count);
+		drop(nominator_count);
 
-		let mut validator_count = 0u32;
-		for (validator, _) in <Validators<T>>::iter() {
+		let max_allowed_len = maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count);
```

Comment:
usually instead of writing drope we use scope like this, it looks better IMO
```
		let max_allowed_len = {
			let nominator_count = CounterForNominators::<T>::get() as usize;
			let validator_count = CounterForValidators::<T>::get() as usize;
			let all_voter_count = validator_count.saturating_add(nominator_count);
			maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count)
		};
```